File: poker/src/client/src/lib/gameState.ts
// Poker Sharp — Game State Management
// All state is in-memory (React state), no localStorage per sandbox constraints

import type { Card } from './poker';
import type { ScoringResult, BoardTexture, GroupedScoringResult } from './scoring';

export interface DrillConfig {
  holdingsCount: number; // 5, 10, 15, 20
  timerEnabled: boolean;
}

export interface DrillRecord {
  id: string;
  date: Date;
  board: Card[];
  targetCount: number;
  timerEnabled: boolean;
  timeTakenMs: number | null;
  hintUsed: boolean;
  userHoldings: { cards: [Card, Card]; rank: number }[];
  scoringResult: ScoringResult | GroupedScoringResult;
  boardTexture?: BoardTexture;
  suitSensitive: boolean; // was this a suit-sensitive board?
}

export interface AppSettings {
  darkMode: boolean;
  playerName: string;
  defaultHoldingsCount: number;
  defaultTimerEnabled: boolean;
}

export const defaultConfig: DrillConfig = {
  holdingsCount: 10,
  timerEnabled: true,
};

export const defaultSettings: AppSettings = {
  darkMode: true, // dark mode is primary for poker feel
  playerName: 'Player',
  defaultHoldingsCount: 10,
  defaultTimerEnabled: true,
};

// Generate a unique ID
export function generateId(): string {
  return Date.now().toString(36) + Math.random().toString(36).slice(2, 7);
}

// Stats calculation helpers
export function getStats(history: DrillRecord[]) {
  const today = new Date();
  today.setHours(0, 0, 0, 0);

  const drillsToday = history.filter(d => {
    const drillDate = new Date(d.date);
    drillDate.setHours(0, 0, 0, 0);
    return drillDate.getTime() === today.getTime();
  }).length;

  // Current streak (consecutive days with at least 1 drill)
  let streak = 0;
  const dayMs = 86400000;
  let checkDate = new Date(today);

  while (true) {
    const hasForDay = history.some(d => {
      const drillDate = new Date(d.date);
      drillDate.setHours(0, 0, 0, 0);
      return drillDate.getTime() === checkDate.getTime();
    });
    if (hasForDay) {
      streak++;
      checkDate = new Date(checkDate.getTime() - dayMs);
    } else {
      break;
    }
  }

  // Longest streak
  let longestStreak = 0;
  if (history.length > 0) {
    const allDates = [...new Set(history.map(d => {
      const dt = new Date(d.date);
      dt.setHours(0, 0, 0, 0);
      return dt.getTime();
    }))].sort((a, b) => a - b);

    let currentRun = 1;
    let maxRun = 1;
    for (let i = 1; i < allDates.length; i++) {
      if (allDates[i] - allDates[i - 1] === dayMs) {
        currentRun++;
        maxRun = Math.max(maxRun, currentRun);
      } else {
        currentRun = 1;
      }
    }
    longestStreak = maxRun;
  }

  // Best accuracy (last 7 days)
  const weekAgo = new Date(today.getTime() - 7 * dayMs);
  const recentDrills = history.filter(d => new Date(d.date) >= weekAgo);
  const bestAccuracy = recentDrills.length > 0
    ? Math.max(...recentDrills.map(d => d.scoringResult.percentage))
    : 0;

  const totalDrills = history.length;

  return { drillsToday, streak, longestStreak, bestAccuracy, totalDrills };
}

/**
 * Calculate accuracy comparison between suit-sensitive and non-suit-sensitive boards.
 * Returns { sensitiveAvg, nonSensitiveAvg, sensitiveCount, nonSensitiveCount }
 * so Stats.tsx can display the comparison.
 */
export function getSuitSensitivity(history: DrillRecord[]) {
  const sensitive = history.filter(d => d.suitSensitive);
  const nonSensitive = history.filter(d => !d.suitSensitive);

  const avg = (records: DrillRecord[]) =>
    records.length === 0
      ? null
      : Math.round(records.reduce((sum, d) => sum + d.scoringResult.percentage, 0) / records.length);

  return {
    sensitiveAvg: avg(sensitive),
    nonSensitiveAvg: avg(nonSensitive),
    sensitiveCount: sensitive.length,
    nonSensitiveCount: nonSensitive.length,
  };
}

File: poker-mobile/lib/core/game_state.dart
// Poker Sharp — Game State Management

import 'poker.dart';
import 'scoring.dart';

enum PickerMode { twoStep, fullGrid }

class DrillConfig {
  final int holdingsCount; // 5, 10, 15, 20
  final bool timerEnabled;
  final PickerMode pickerMode;

  const DrillConfig({
    this.holdingsCount = 10,
    this.timerEnabled = true,
    this.pickerMode = PickerMode.twoStep,
  });

  DrillConfig copyWith({
    int? holdingsCount,
    bool? timerEnabled,
    PickerMode? pickerMode,
  }) {
    return DrillConfig(
      holdingsCount: holdingsCount ?? this.holdingsCount,
      timerEnabled: timerEnabled ?? this.timerEnabled,
      pickerMode: pickerMode ?? this.pickerMode,
    );
  }
}

class DrillRecord {
  final String id;
  final DateTime date;
  final List<Card> board;
  final int targetCount;
  final PickerMode pickerMode;
  final bool timerEnabled;
  final int? timeTakenMs;
  final bool hintUsed;
  final List<({List<Card> cards, int rank})> userHoldings;
  final GroupedScoringResult? scoringResult;
  final BoardTexture? boardTexture;
  final bool suitSensitive;

  const DrillRecord({
    required this.id,
    required this.date,
    required this.board,
    required this.targetCount,
    required this.pickerMode,
    required this.timerEnabled,
    this.timeTakenMs,
    this.hintUsed = false,
    this.userHoldings = const [],
    this.scoringResult,
    this.boardTexture,
    this.suitSensitive = false,
  });
}

class AppSettings {
  final bool darkMode;
  final String playerName;
  final int defaultHoldingsCount;
  final bool defaultTimerEnabled;
  final PickerMode defaultPickerMode;
  final bool soundEnabled;
  final bool hapticEnabled;

  const AppSettings({
    this.darkMode = true,
    this.playerName = 'Player',
    this.defaultHoldingsCount = 10,
    this.defaultTimerEnabled = true,
    this.defaultPickerMode = PickerMode.twoStep,
    this.soundEnabled = true,
    this.hapticEnabled = true,
  });

  AppSettings copyWith({
    bool? darkMode,
    String? playerName,
    int? defaultHoldingsCount,
    bool? defaultTimerEnabled,
    PickerMode? defaultPickerMode,
    bool? soundEnabled,
    bool? hapticEnabled,
  }) {
    return AppSettings(
      darkMode: darkMode ?? this.darkMode,
      playerName: playerName ?? this.playerName,
      defaultHoldingsCount: defaultHoldingsCount ?? this.defaultHoldingsCount,
      defaultTimerEnabled: defaultTimerEnabled ?? this.defaultTimerEnabled,
      defaultPickerMode: defaultPickerMode ?? this.defaultPickerMode,
      soundEnabled: soundEnabled ?? this.soundEnabled,
      hapticEnabled: hapticEnabled ?? this.hapticEnabled,
    );
  }
}

const defaultConfig = DrillConfig();
const defaultSettings = AppSettings();

/// Generate a unique ID.
String generateId() {
  final now = DateTime.now().millisecondsSinceEpoch;
  final random = now.toRadixString(36) +
      (DateTime.now().microsecond).toRadixString(36).padLeft(4, '0');
  return random;
}

/// Stats calculation from drill history.
({
  int drillsToday,
  int streak,
  int longestStreak,
  int bestAccuracy,
  int totalDrills,
}) getStats(List<DrillRecord> history) {
  final now = DateTime.now();
  final today = DateTime(now.year, now.month, now.day);

  final drillsToday = history.where((d) {
    final drillDate = DateTime(d.date.year, d.date.month, d.date.day);
    return drillDate == today;
  }).length;

  // Current streak (consecutive days with at least 1 drill)
  int streak = 0;
  const dayMs = 86400000;
  var checkDate = today;

  while (true) {
    final hasForDay = history.any((d) {
      final drillDate = DateTime(d.date.year, d.date.month, d.date.day);
      return drillDate == checkDate;
    });
    if (hasForDay) {
      streak++;
      checkDate = checkDate.subtract(const Duration(days: 1));
    } else {
      break;
    }
  }

  // Longest streak
  int longestStreak = 0;
  if (history.isNotEmpty) {
    final allDates = history
        .map((d) => DateTime(d.date.year, d.date.month, d.date.day)
            .millisecondsSinceEpoch)
        .toSet()
        .toList()
      ..sort();

    int currentRun = 1;
    int maxRun = 1;
    for (int i = 1; i < allDates.length; i++) {
      if (allDates[i] - allDates[i - 1] == dayMs) {
        currentRun++;
        if (currentRun > maxRun) maxRun = currentRun;
      } else {
        currentRun = 1;
      }
    }
    longestStreak = maxRun;
  }

  // Best accuracy (last 7 days)
  final weekAgo = today.subtract(const Duration(days: 7));
  final recentDrills = history.where((d) => d.date.isAfter(weekAgo)).toList();
  final bestAccuracy = recentDrills.isNotEmpty
      ? recentDrills
          .map((d) => d.scoringResult?.percentage ?? 0)
          .reduce((a, b) => a > b ? a : b)
      : 0;

  final totalDrills = history.length;

  return (
    drillsToday: drillsToday,
    streak: streak,
    longestStreak: longestStreak,
    bestAccuracy: bestAccuracy,
    totalDrills: totalDrills,
  );
}

/// Calculate accuracy comparison between suit-sensitive and non-suit-sensitive boards.
({
  int? sensitiveAvg,
  int? nonSensitiveAvg,
  int sensitiveCount,
  int nonSensitiveCount,
}) getSuitSensitivity(List<DrillRecord> history) {
  final sensitive = history.where((d) => d.suitSensitive).toList();
  final nonSensitive = history.where((d) => !d.suitSensitive).toList();

  int? avg(List<DrillRecord> records) {
    if (records.isEmpty) return null;
    final sum = records.fold<int>(
        0, (acc, d) => acc + (d.scoringResult?.percentage ?? 0));
    return (sum / records.length).round();
  }

  return (
    sensitiveAvg: avg(sensitive),
    nonSensitiveAvg: avg(nonSensitive),
    sensitiveCount: sensitive.length,
    nonSensitiveCount: nonSensitive.length,
  );
}
